### PR TITLE
feat(staking): adds e2e test runner

### PR DESCRIPTION
**Pull request by damirka**
Creates a simple wrapper around `Scenario` to init the application and run transactions in a better manner. 

```rust
// create and configure objects
let admin = @0xa11ce
let mut runner = e2e_runner::prepare(admin)
    .epoch_zero_duration(EPOCH_ZERO_DURATION) // optionally change parameters
    .n_shards(N_SHARDS)
    .build();

// access clock
runner.clock().increment_for_testing(EPOCH_ZERO_DURATION);

// ...
// run tx
runner.tx!(admin, |staking, system, _ctx| {
    staking.voting_end(runner.clock());
    staking.initiate_epoch_change(system, runner.clock());

    assert!(system.epoch() == 1);
    assert!(system.committee().n_shards() == N_SHARDS);

    nodes.do_ref!(|node| assert!(system.committee().contains(&node.node_id())));
});

```

### CODE REVIEW

File: contracts/walrus/sources/staking/active_set.move [hunk 1]
```diff
@@ -124,9 +124,9 @@ fun insert_impl(set: &mut ActiveSet, node_id: ID, staked_amount: u64) {
     set.nodes.insert(node_id, staked_amount);
     let map_idx = set.nodes.size() as u16 - 1;
     let insert_idx = set.idx_sorted.find_index!(|idx| {
-            let (_node_id, stake) = set.nodes.get_entry_by_idx(*idx as u64);
-            staked_amount > *stake
-        }).destroy_or!(set.idx_sorted.length());
+        let (_node_id, stake) = set.nodes.get_entry_by_idx(*idx as u64);
+        staked_amount > *stake
+    }).destroy_or!(set.idx_sorted.length());
     set.idx_sorted.insert(insert_idx as u16, map_idx as u64);
```
Comment (gdanezis) [inline contracts/walrus/sources/staking/active_set.move:130]: The above code I think does a linear search (?) to find a position, and the there is a linear cost operation to insert. Is this ok for largish committees?

File: contracts/walrus/sources/system/bls_aggregate.move [hunk 2]
```diff
@@ -69,7 +69,7 @@ public(package) fun new_bls_committee_member(
 // === Accessors for BlsCommittee ===
 
 /// Get the epoch of the committee.
-public fun epoch(self: &BlsCommittee): u32 {
+public(package) fun epoch(self: &BlsCommittee): u32 {
```
Comment (gdanezis) [inline contracts/walrus/sources/system/bls_aggregate.move:72]: Is there a reason to restrict? I am usually live and let live with read accessors - but up to you.

File: contracts/walrus/tests/e2e_runner.move [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+/// Take any shared resource and call the function `f` with it and the `TxContext` as arguments.
+public macro fun shared_tx<$T>(
+    $runner: &mut TestRunner,
+    $sender: address,
+    $f: |&mut $T, &mut TxContext|,
+) {
+    let runner = $runner;
+    let scenario = runner.scenario();
+    scenario.next_tx($sender);
+    let mut staking = scenario.take_shared<$T>();
```
Comment (gdanezis) [inline contracts/walrus/tests/e2e_runner.move]: I would change the var name away from `staking` to `shared_object` or something like that. This is not always a staking object unlike the above function.

File: contracts/walrus/tests/e2e_runner.move [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    let ctx = scenario.ctx();
+
+    $f(&mut staking, &mut system, ctx);
+
+    test_scenario::return_shared(staking);
+    test_scenario::return_shared(system);
+}
+
+/// Take any shared resource and call the function `f` with it and the `TxContext` as arguments.
+public macro fun shared_tx<$T>(
```
Comment (gdanezis) [inline contracts/walrus/tests/e2e_runner.move]: Maybe call this `tx_with_any_shared`?

File: contracts/walrus/tests/e2e_tests.move [hunk 5]
```diff
@@ -55,127 +39,97 @@ fun test_init_and_first_epoch_change(): (staking::Staking, system::System) {
...
-    {
-        staking.voting_end(&clock);
-        staking.initiate_epoch_change(&mut system, &clock);
-    };
+runner.clock().increment_for_testing(EPOCH_ZERO_DURATION);
+runner.tx!(admin, |staking, system, _| {
+    staking.voting_end(runner.clock());
+    staking.initiate_epoch_change(system, runner.clock());
 
-    // === check if epoch was changed as expected ===
```
Comment (gdanezis) [inline contracts/walrus/tests/e2e_tests.move:92]: Maybe keep the comment?

File: contracts/walrus/tests/e2e_tests.move [hunk 6]
```diff
@@ -55,127 +39,97 @@ fun test_init_and_first_epoch_change(): (staking::Staking, system::System) {
...
-    clock.increment_for_testing(EPOCH_DURATION - PARAM_SELECTION_DELTA);
-    scenario.next_tx(nodes[0].sui_address());
-    {
-        staking.initiate_epoch_change(&mut system, &clock);
-    };
+    runner.clock().increment_for_testing(EPOCH_DURATION - PARAM_SELECTION_DELTA);
+    runner.tx!(admin, |staking, system, _| {
+        staking.initiate_epoch_change(system, runner.clock());
 
-    // === check if epoch was changed as expected ===
```
Comment (gdanezis) [inline contracts/walrus/tests/e2e_tests.move:124]: Nothing wrong with this comment, why not keep it.

File: contracts/walrus/tests/test_utils.move [hunk 7]
```diff
@@ -57,7 +57,8 @@ public struct PoolBuilder has copy, drop {
 /// let pool_c = pool()
 ///     .name(b"my node".to_string())
 ///     .network_address(b"0.0.0.0".to_string())
-///     .public_key(x"a60e75190e62b6a54142d147289a735c4ce11a9d997543da539a3db57def5ed83ba40b74e55065f02b35aa1d504c404b")
+///
+/// .public_key(x"a60e75190e62b6a54142d147289a735c4ce11a9d997543da539a3db57def5ed83ba40b74e55065f02b35aa1d504c404b")
```
Comment (gdanezis) [inline contracts/walrus/tests/test_utils.move]: Keep the indent here? And remove space?

File: contracts/walrus/tests/e2e_runner.move (hunk 4 above)
Comment (damirka) [reply contracts/walrus/tests/e2e_runner.move]: Oh, this buddy needs to go away. It's a leftover.

File: contracts/walrus/tests/e2e_tests.move (hunk 6 above)
Comment (damirka) [reply contracts/walrus/tests/e2e_tests.move:124]: The comment is still there, just that two operations are now merged into one.

File: contracts/walrus/sources/system/bls_aggregate.move (hunk 2 above)
Comment (damirka) [reply contracts/walrus/sources/system/bls_aggregate.move:72]: This method stood out, the rest are `public(package)`. I'd like to keep all of them private to the package until we decide on which APIs to keep public.

File: contracts/walrus/sources/staking/active_set.move (hunk 1 above)
Comment (damirka) [reply contracts/walrus/sources/staking/active_set.move:130]: This is purely a formatting change, I believe @tnowacki was or is looking into that.

File: contracts/walrus/sources/staking/active_set.move [hunk 8]
```diff
@@ -124,9 +124,9 @@ fun insert_impl(set: &mut ActiveSet, node_id: ID, staked_amount: u64) {
     set.nodes.insert(node_id, staked_amount);
     let map_idx = set.nodes.size() as u16 - 1;
     let insert_idx = set.idx_sorted.find_index!(|idx| {
-            let (_node_id, stake) = set.nodes.get_entry_by_idx(*idx as u64);
-            staked_amount > *stake
-        }).destroy_or!(set.idx_sorted.length());
+        let (_node_id, stake) = set.nodes.get_entry_by_idx(*idx as u64);
+        staked_amount > *stake
+    }).destroy_or!(set.idx_sorted.length());
```
Comment (tnowacki) [inline contracts/walrus/sources/staking/active_set.move:129]: This auto formatted? No idea how to feel about this sort of case

File: contracts/walrus/sources/system/bls_aggregate.move [hunk 9]
```diff
@@ -78,6 +78,11 @@ public(package) fun n_shards(self: &BlsCommittee): u16 {
     self.n_shards
 }
 
+/// Checks if the committee contains a given node.
+public(package) fun contains(self: &BlsCommittee, node_id: &ID): bool {
+    self.members.find_index!(|member| member.node_id == *node_id).is_some()
```
Comment (tnowacki) [inline contracts/walrus/sources/system/bls_aggregate.move:83]:
```
&member.node_id == node_id
```
is conceptually cheaper, even if not in terms of gas costs

**Pull request merged**